### PR TITLE
NAS-121177 / 23.10 / Refactor IxSlideInService

### PR DESCRIPTION
**Pull request by Raz-Dva**
Method open of IxSlideInService returns 
      .componentInstance
      .close()
      .afterClosed$

It is possible to pass data to the created component using
@Inject(SLIDEIN_DATA) data

### CODE REVIEW

File: src/app/modules/ix-forms/components/ix-slide-in/ix-slide-in.component.ts [hunk 1]
```diff
@@ -87,7 +88,9 @@ export class IxSlideInComponent implements OnInit, OnDestroy {
     this.slideInBody.clear();
     this.wasBodyCleared = false;
 
-    const componentRef = this.slideInBody.createComponent<T>(componentType);
+    const injector = Injector.create({ providers: [{ provide: SLIDEIN_DATA, useValue: params?.data }] });
```
Comment (undsoft) [inline src/app/modules/ix-forms/components/ix-slide-in/ix-slide-in.component.ts:91]: SLIDE_IN_DATA

File: src/app/services/ix-slide-in.service.ts [hunk 2]
```diff
@@ -32,9 +40,13 @@ export class IxSlideInService {
...
 
-  open<T>(modal: Type<T>, params?: { wide: boolean }): T {
+  open<T>(modal: Type<T>, params?: { wide: boolean; data?: { [key: string]: unknown } }): SlideInServiceRef<T> {
     this.modalType = modal;
-    return this.slideInComponent.openSlideIn(modal, params);
+    return {
+      componentInstance: this.slideInComponent.openSlideIn(modal, params),
+      close: this.close,
+      afterClosed$: this.onClose$,
+    };
```
Comment (undsoft) [inline src/app/services/ix-slide-in.service.ts:49]: The idea is to be able to tell when a specific slide in has been closed.
For example, in material dialogs I can open individual dialog, tell when it was closed and receive data from it.
```
this.dialog.open(MyDialog)
  .afterClosed()
  .subscribe((result) => {
    console.log('My Dialog was closed with result', result);
  });
```

This is not possible right now. `onClose$` will emit for every slide in closed, not just the one I'm opening.

**Review by undsoft — CHANGES_REQUESTED** [review]
You didn't have to update every usage just yet. It's important to get API right before updating every place.
You can just create `open2` method temporarily for example and update usage in one place or whatever is easier.

File: src/app/modules/ix-forms/components/ix-slide-in/ix-slide-in-ref.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { Type } from '@angular/core';
+import { Observable, Subject } from 'rxjs';
+import { IxSlideInComponent } from 'app/modules/ix-forms/components/ix-slide-in/ix-slide-in.component';
+import { ResponseOnClose } from 'app/services/ix-slide-in.service';
+
+export class IxSlideInRef<T> {
```
Comment (undsoft) [inline src/app/modules/ix-forms/components/ix-slide-in/ix-slide-in-ref.ts:6]: Let's also add second generic R for returning value.

File: src/app/modules/ix-forms/components/ix-slide-in/ix-slide-in-ref.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  private slideInClosed$ = new Subject<ResponseOnClose>();
+  componentInstance: T;
+  uuid: string;
+  constructor(
+    private slideInComponent: IxSlideInComponent,
+    private modalType: Type<T>,
+  ) {}
+
+  closeThisSlide(error?: Error, response?: unknown): void {
```
Comment (undsoft) [inline src/app/modules/ix-forms/components/ix-slide-in/ix-slide-in-ref.ts:15]: Let's rename to just `close` and change arguments to `response: R`. I don't think we need to pass modalType or error. What to return would be up to developer.

File: src/app/services/ix-slide-in.service.ts [hunk 5]
```diff
@@ -32,11 +35,13 @@ export class IxSlideInService {
     this.slideInComponent = modal;
   }
 
-  open<T>(modal: Type<T>, params?: { wide: boolean }): T {
+  open<T>(modal: Type<T>, params?: { wide?: boolean; data?: { [key: string]: unknown } }): IxSlideInRef<T> {
```
Comment (undsoft) [inline src/app/services/ix-slide-in.service.ts:38]: Let's also add a generic for returning value, defaulting to `unknown`.

File: src/app/pages/data-protection/components/data-protection-dashboard/data-protection-dashboard.component.ts [hunk 6]
```diff
@@ -115,7 +118,7 @@ export class DataProtectionDashboardComponent implements OnInit {
 
     merge(
       this.modalService.onClose$,
-      this.slideInService.onClose$,
+      this.afterClosedSlideIn$.pipe(mergeMap((afterClosed$) => afterClosed$)),
```
Comment (undsoft) [inline src/app/pages/data-protection/components/data-protection-dashboard/data-protection-dashboard.component.ts:121]: This is perhaps a bad example for this. I think in the future, this component would be split up into smaller card components and each card will responsible for reloading itself once slideIn is closed as opposed to having this method.

**Review by undsoft — APPROVED** [review]
Looks good. You can either update usages everywhere or keep old methods and add new methods and do replacements in future PRs.

**Comment by Raz-Dva** [comment]
For the working pull request, I will create a new ixSlideService.
I will make pull requests gradually removing the methods of the old service.

**Comment by undsoft** [comment]
@Raz-Dva Can this PR be closed since you have another one?

**Comment by Raz-Dva** [comment]
Another one is open for this issue  https://github.com/truenas/webui/pull/8120

**Pull request closed without merging**